packages/apollo-server-plugin-response-cache/src/ApolloServerPluginResponseCache.ts

```diff
@@ -0,0 +1,337 @@
+  // the request other than the query document, operation name, variables, and
+  // session ID. For example, responses that include translatable text may want
+  // to return a string derived from
+  // requestContext.request.http.headers.get('Accept-Language'). The data may
+  // be anything that can be JSON-stringified.
+  extraCacheKeyData?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<any>;
+
+  // If this hook is defined and returns false, the plugin will not read
+  // responses from the cache.
+  shouldReadFromCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+
+  // If this hook is defined and returns false, the plugin will not write the
+  // response to the cache.
+  shouldWriteToCache?(
+    requestContext: GraphQLRequestContext<TContext>,
+  ): ValueOrPromise<boolean>;
+}
+
+enum SessionMode {
+  NoSession,
+  Private,
+  AuthenticatedPublic,
+}
+
+function sha(s: string) {
+  return createHash('sha256')
+    .update(s)
+    .digest('hex');
+}
+
+interface BaseCacheKey {
+  documentText: string;
+  operationName: string | null;
+  variables: { [name: string]: any };
+  extra: any;
+}
+
+interface ContextualCacheKey {
+  sessionMode: SessionMode;
+  sessionId?: string | null;
+}
+
+interface CacheValue {
+  // Note: we only store data responses in the cache, not errors.
+  //
+  // There are two reasons we don't cache errors. The user-level reason is that
+  // we think that in general errors are less cacheable than real results, since
+  // they might indicate something transient like a failure to talk to a
+  // backend. (If you need errors to be cacheable, represent the erroneous
+  // condition explicitly in data instead of out-of-band as an error.) The
+  // implementation reason is that this lets us avoid complexities around
+  // serialization and deserialization of GraphQL errors, and the distinction
+  // between formatted and unformatted errors, etc.
+  data: Record<string, any>;
+  cachePolicy: Required<CacheHint>;
+  cacheTime: number; // epoch millis, used to calculate Age header
+}
+
+type CacheKey = BaseCacheKey & ContextualCacheKey;
+
+function cacheKeyString(key: CacheKey) {
+  return sha(JSON.stringify(key));
+}
+
+function isGraphQLQuery(requestContext: GraphQLRequestContext<any>) {
+  return requestContext.operation!.operation === 'query';
+}
+
+export default function plugin(
+  options: Options = Object.create(null),
+): ApolloServerPlugin {
+  return {
+    requestDidStart(
+      outerRequestContext: GraphQLRequestContext<any>,
+    ): GraphQLRequestListener<any> {
+      const cache = new PrefixingKeyValueCache(
+        options.cache || outerRequestContext.cache!,
+        'fqc:',
+      );
+
+      let sessionId: string | null = null;
+      let baseCacheKey: BaseCacheKey | null = null;
+      let age: number | null = null;
+
+      return {
+        async responseForOperation(
+          requestContext: WithRequired<
```

Comment:
Hmm, this should be already inferred from the `apollo-server-plugin-base` and not necessary to be explicitly defined here, no?

Reply:
What do you mean — when implementing a method you don't need to declare its arguments?

Reply:
I believe that you can just have `requestContext` here (leaving off the duplicative re-typing) and the type will be inferred as the `WithRequired<....>` type as defined in [this bit from above](https://github.com/apollographql/apollo-server/pull/2437/files#diff-0f0cc848fe2c6442424f2ea0559f35ecR40)?